Added construct workflow documentation (#85)
litmuschaos/litmus-docs: website/docs/user-guides/construct-workflow.md

sidebar_label: Construct Workflow YAML
---

**Chaos Workflow** is a set of different operations coupled together to achieve desired chaos imapact on a Kubernetes Cluster. <br/>
A basic chaos workflow consists of these steps: 
1. Install ChaosExperiment
2. Install ChaosEngine
3. Revert Chaos

## Before we begin

> To construct a Chaos Workflow without ChaosCenter, make sure you are aware of [Chaos Workflow](../concepts/chaos-workflow) and the different steps present in it.  

Reviewer: We should add Chaos Engine too as a Concept they should know before beginning

## Steps to Construct a Chaos Workflow

> LitmusChaos leverages the popular workflow and GitOps tool **Argo** to achieve this goal. Argo enables the creation of different chaos scenarios together in from of workflows which are extremly simple and efficient to use.<br/>

Reviewer 2: ```suggestion
LitmusChaos leverages the popular workflow and GitOps tool **Argo** to achieve this goal. Argo enables the orchestrating of different chaos scenarios together in form of workflow which is extremely simple and efficient to use.<br/> 
```

Reviewer: *orchestration

Reviewer: *in the form of

Here is a sample pod-delete chaos workflow.

> ```yaml

Reviewer: can we please add embedmd to always stay in sync with the updated yaml!

                  jobCleanUpPolicy: retain
                  engineState: active
                  chaosServiceAccount: litmus-admin
                  experiments:
                    - name: pod-delete
                      spec:
                        components:
                          env:
                            - name: TOTAL_CHAOS_DURATION
                              value: "30"
                            - name: CHAOS_INTERVAL
                              value: "10"
                            - name: FORCE
                              value: "false"
                            - name: PODS_AFFECTED_PERC
                              value: ""
      container:
        args:
          - -file=/tmp/chaosengine-pod-delete.yaml
          - -saveName=/tmp/engine-name
        image: litmuschaos/litmus-checker:latest
    - name: revert-chaos
      container:
        image: litmuschaos/k8s:latest
        command:
          - sh
          - -c
        args:
          - "kubectl delete chaosengine -l 'instance_id in
            (86a4f130-d99b-4e91-b34b-8f9eee22cb63, )' -n
            {{workflow.parameters.adminModeNamespace}} "
```

> The structure of a chaos workflow is similar to that of a Kubernetes Object. It consists of the mandatory fields like `apiVersion`, `kind`, `metadata`, `spec`.

Reviewer 2: Can we please explain its components and after that workflow example?

Few additional terms in an Argo workflows are:
1. **Template** : It consists of different steps with their specific operations.
```yaml
      templates:
    - name: custom-chaos
      steps:
        - - name: install-chaos-experiments
            template: install-chaos-experiments
        - - name: pod-delete
            template: pod-delete
        - - name: revert-chaos

>             template: revert-chaos

Reviewer 2: Will you please add one more example for parallel steps?

## Resources
- The ChaosExperiment CR and ChaosEngine CR of different experiments are available at **[ChaosHub](http://hub.litmuschaos.io/)**.

<iframe width="560" height="315" src="https://www.youtube.com/embed/dan0AXO_soY?start=310" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

<iframe width="560" height="315" src="https://www.youtube.com/embed/xLjTx8lqTuQ?start=1163" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

> ## Learn More

Reviewer 2: Add experiment and results also

> - [What is a Chaos Workflow](../concepts/chaos-workflow)

Reviewer: Chaos Workflow and Chaos Engine are already covered as concepts the user needs to know thus you should put other concepts and user-guides relevant to this topic other than those.

> - [What is a Chaos Workflow](../concepts/chaos-engine)

Reviewer 2: ```suggestion
- [What is a Chaos Engine](../concepts/chaos-engine)
```
